### PR TITLE
Add GetLeavesByRange entrypoint

### PR DESCRIPTION
**Pull request by daviddrysdale**

### CODE REVIEW

File: trillian_log_api.proto [hunk 1]
```diff
@@ -154,6 +154,19 @@ message GetLeavesByIndexResponse {
...
+message GetLeavesByRangeRequest {
+  int64 log_id = 1;
+  int64 start_index = 2;
+  int64 count = 3;
+}
+
+message GetLeavesByRangeResponse {
+  // Returned log leaves starting from the start_index of the request, in order.
+  // There may be fewer than request.count leaves returned, if the requested
+  // range extended beyond the size of the tree.
```
Comment (AlCutter) [inline trillian_log_api.proto]: Or if the server opted to return fewer than reqeuested...

File: client/client.go [hunk 2]
```diff
@@ -88,40 +87,28 @@ func (c *LogClient) GetByIndex(ctx context.Context, index int64) (*trillian.LogL
...
-	if got, want := len(resp.Leaves), len(indexes); got != want {
-		return nil, fmt.Errorf("len(Leaves): %v, want %v", got, want)
+	if len(resp.Leaves) < int(count) {
+		return nil, fmt.Errorf("len(Leaves)=%d, want %d", len(resp.Leaves), count)
 	}
 	for i, l := range resp.Leaves {
-		if got, want := l.LeafIndex, indexes[i]; got != want {
-			return nil, fmt.Errorf("Leaves[%v].Index: %v, want %v", i, got, want)
+		if l.LeafIndex != start+int64(i) {
+			return nil, fmt.Errorf("Leaves[%d].LeafIndex=%d, want %d", i, l.LeafIndex, count+int64(i))
```
Comment (AlCutter) [inline client/client.go]: `start+int64(i)` in the error too, no? 
this is why it's a good idea to use the `if x, y := ..., ...; x!=y { log(..., x, y) }` pattern!

File: trillian_log_api.proto (hunk 1 above)
Comment (daviddrysdale) [reply trillian_log_api.proto]: Done.

File: client/client.go (hunk 2 above)
Comment (daviddrysdale) [reply client/client.go]: Done.

**Comment by pavelkalinnikov** [comment]
A belated comment on this. Google Cloud API guide [suggests](https://cloud.google.com/apis/design/design_patterns#representing_ranges) to avoid range representations other than `[begin, end)` kind of thing. Is it too late to change that?

**Pull request merged**